pkg/checks/traceroute/traceroute.go

```diff
@@ -2,197 +2,263 @@ package traceroute
+				log.Error("traceroute could not reach target", "ttl", ttl)
 			}
+		}()
+	}
 
-			l.Debug("Ran traceroute", "result", trace, "duration", duration)
+	wg.Wait()
+	close(results)
 
-			r := result{
-				NumHops: len(trace.Hops),
-				Hops:    []hop{},
-			}
+	for r := range results {
+		hops[r.Ttl] = append(hops[r.Ttl], r)
+	}
 
-			for _, h := range trace.Hops {
-				r.Hops = append(r.Hops, hop{
-					Addr:    h.Host,
-					Latency: h.ElapsedTime,
-					Success: h.Success,
-				})
-			}
-			cResult <- internalResult{addr: t.Addr, res: r}
-		}(t)
+	log.Debug("finished trace", "hops", printHops(hops))
+
+	return hops, nil
+}
+
+func ipFromAddr(remoteAddr net.Addr) net.IP {
+	switch addr := remoteAddr.(type) {
+	case *net.UDPAddr:
+		return addr.IP
+	case *net.TCPAddr:
+		return addr.IP
+	case *net.IPAddr:
+		return addr.IP
 	}
+	return nil
+}
 
-	log.Debug("Waiting for traceroute checks to finish")
+func traceroute(results chan Hop, addr net.Addr, ttl int, timeout time.Duration) (bool, error) {
+	canIcmp := true
+	icmpListener, err := icmp.ListenPacket("ip4:icmp", "0.0.0.0")
+	if err != nil {
+		if !errors.Is(err, syscall.EPERM) {
+			return false, err
+		}
+		canIcmp = false
+	}
 
-	go func() {
-		wg.Wait()
-		close(cResult)
+	defer func() {
+		if canIcmp {
+			icmpListener.Close() // #nosec G104
+		}
 	}()
+	start := time.Now()
+	conn, clientPort, err := tcpHop(addr, ttl, timeout)
+	latency := time.Since(start)
+	if err == nil {
+		conn.Close() // #nosec G104
+
+		ipaddr := ipFromAddr(addr)
+		names, _ := net.LookupAddr(ipaddr.String()) // we don't care about this lookup failling
+
+		name := ""
+		if len(names) >= 1 {
+			name = names[0]
+		}
 
-	log.Debug("All traceroute checks finished")
-
-	for r := range cResult {
-		res[r.addr] = r.res
+		results <- Hop{
+			Latency: latency,
+			Ttl:     ttl,
+			Addr:    addr,
+			Name:    name,
+			Reached: true,
+		}
+		return true, nil
 	}
 
-	return res
-}
+	found := false
+	deadline := time.Now().Add(timeout)
 
-// Shutdown is called once when the check is unregistered or sparrow shuts down
-func (tr *Traceroute) Shutdown() {
-	tr.DoneChan <- struct{}{}
-	close(tr.DoneChan)
-}
+	for time.Now().Unix() < deadline.Unix() && !found {
+		gotPort := -1
+		var addr net.Addr
+		if canIcmp {
+			gotPort, addr, err = readIcmpMessage(icmpListener, timeout)
+			if err != nil {
+				results <- Hop{
+					Latency: latency,
+					Ttl:     ttl,
+					Reached: false,
+				}
+				return false, nil
+			}
+		}
 
-// SetConfig is called once when the check is registered
-// This is also called while the check is running, if the remote config is updated
-// This should return an error if the config is invalid
-func (tr *Traceroute) SetConfig(cfg checks.Runtime) error {
-	if cfg, ok := cfg.(*Config); ok {
-		tr.Mu.Lock()
-		defer tr.Mu.Unlock()
-		tr.config = *cfg
-		return nil
-	}
+		// Check if the destination port matches our dialer's source port
+		if canIcmp && gotPort == clientPort {
```

Comment:
You could simplify this by using:

```golang
if !canIcmp {
  break
}
```

Reply:
I think this func should be slitted, its hard to read and to reduce complexity. Maybe a func `watchForIcmp` (or sth else) can be used that is called when initial tcpHop has failed